pkg/placement/placement.go

```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
-func (p *Service) PerformTablesUpdate(hosts []placementv1pb.Placement_ReportDaprStatusServer,
-	options placementOptions) {
-	p.updateLock.Lock()
-	defer p.updateLock.Unlock()
-
-	if options.incrementGeneration {
-		p.generation++
-	}
-
-	o := placementv1pb.PlacementOrder{
-		Operation: "lock",
-	}
-
-	for _, host := range hosts {
-		err := host.Send(&o)
-		if err != nil {
-			log.Errorf("error updating host on lock operation: %s", err)
-			continue
-		}
-	}
-
-	v := fmt.Sprintf("%v", p.generation)
-
-	o.Operation = "update"
-	o.Tables = &placementv1pb.PlacementTables{
-		Version: v,
-		Entries: map[string]*placementv1pb.PlacementTable{},
-	}
-
-	for k, v := range p.entries {
-		hosts, sortedSet, loadMap, totalLoad := v.GetInternals()
-		table := placementv1pb.PlacementTable{
-			Hosts:     hosts,
-			SortedSet: sortedSet,
-			TotalLoad: totalLoad,
-			LoadMap:   make(map[string]*placementv1pb.Host),
-		}
-
-		for lk, lv := range loadMap {
-			h := placementv1pb.Host{
-				Name: lv.Name,
-				Load: lv.Load,
-				Port: lv.Port,
-				Id:   lv.AppID,
-			}
-			table.LoadMap[lk] = &h
-		}
-		o.Tables.Entries[k] = &table
-	}
-
-	for _, host := range hosts {
-		err := host.Send(&o)
-		if err != nil {
-			log.Errorf("error updating host on update operation: %s", err)
-			continue
-		}
-	}
-
-	o.Tables = nil
-	o.Operation = "unlock"
-
-	for _, host := range hosts {
-		err := host.Send(&o)
-		if err != nil {
-			log.Errorf("error updating host on unlock operation: %s", err)
-			continue
-		}
-	}
-}
-
-// ProcessRemovedHost removes a host from the hash table
-func (p *Service) ProcessRemovedHost(id string) {
-	updateRequired := false
-
-	var entities []string
-	func() {
-		p.hostsEntitiesLock.RLock()
-		defer p.hostsEntitiesLock.RUnlock()
-		entities = p.hostsEntities[id]
-		delete(p.hostsEntities, id)
-	}()
-
-	func() {
-		p.entriesLock.Lock()
-		defer p.entriesLock.Unlock()
-		for _, e := range entities {
-			if _, ok := p.entries[e]; ok {
-				p.entries[e].Remove(id)
-				updateRequired = true
-			}
-		}
-	}()
-
-	if updateRequired {
-		p.PerformTablesUpdate(p.hosts, placementOptions{incrementGeneration: true})
-	}
+func (p *Service) addRuntimeConnection(ctx context.Context, conn placementGRPCStream) {
+	p.streamConnsLock.Lock()
+	p.streamConns = append(p.streamConns, conn)
+	p.streamConnsLock.Unlock()
 }
 
-// ProcessHost updates the distributed has list based on a new host and its entities
-func (p *Service) ProcessHost(host *placementv1pb.Host) {
-	updateRequired := false
-
-	for _, e := range host.Entities {
-		p.entriesLock.Lock()
-
-		if _, ok := p.entries[e]; !ok {
-			p.entries[e] = hashing.NewConsistentHash()
-		}
-
-		exists := p.entries[e].Add(host.Name, host.Id, host.Port)
-		if !exists {
-			updateRequired = true
-			monitoring.RecordPerActorTypeReplicasCount(e, host.Name)
+func (p *Service) deleteRuntimeConnection(conn placementGRPCStream) {
+	p.streamConnsLock.Lock()
+	for i, c := range p.streamConns {
```

Comment:
should streamConns be a map for faster lookup? Since there might be the scenario that there are multiple connections being added and removed for actors frequently.